### PR TITLE
Essay Submission: Comparing DevOps Tools You Will Likely Encounter in the Workplace

### PR DESCRIPTION
**Pull request by ayubatif**
# Comparing DevOps Tools You Will Likely Encounter in the Workplace

## Members

Name: Ayub Atif
Mail: aatif@kth.se
Github: [ayubatif](https://github.com/ayubatif)

Name: Justin Arieltan
Mail: arieltan@kth.se
Github: [Agriad](https://github.com/Agriad)

## Initial Proposal
#1020 

## Feedback

We receive feedback on the essay from @fwallb 

## Submission 

the essay can be found at /contributions/essay/aatif-arieltan/atif_arieltan_Comparing_DevOps_Tools_You_Will_Likely_Encounter_in_the_Workplace.pdf

# WORD COUNT: 2010 words
pdf submission words counted using 
```bash
pdftotext $file - | wc -w
```

## Grading Criteria we aim for
|                                             | Yes | No | Remarkable |
|-------------------------------------------- | ----|----|-------------|
| format: The essay is in PDF | Mandatory🆗 | - | - |
| title: The essay has a good title | Yes🆗 | No | Remarkable🆗 |
| well-structured: The essay is well structured, the ideas are well presented | Yes🆗 | No | Remarkable🆗 |
| introduction: The essay contains a good introduction | Yes🆗 | No | Remarkable |
| conclusion: The essay contains a good conclusion | Yes🆗 | No | Remarkable |
| self-contained: The essay is self-contained, one can understand it without reading something else (expected knowledge of the reader: a master student in computer science) | Yes🆗 | No | Remarkable🆗 |
| innovative: The essay contains innovative ideas or material | Yes | No🆗 | Remarkable |
| figures: The essay contains good figures or listings | Yes🆗 | No | Remarkable |
| sound: The essay is sound, factual, and accurate | Yes🆗 | No | Remarkable🆗 |
| references: The essay contains references, appropriate  in number and quality (10 good refs is a minimum) | Yes🆗 | No | Remarkable🆗 |
| elegant: The essay presentation is elegant / visually appealing (eg LaTeX, InDesign) | Yes🆗 | No | Remarkable🆗 |
| relevant: The essay addresses a topic that is relevant for DevOps | Yes🆗 | No | Remarkable🆗 |

### CODE REVIEW

**Comment by ayubatif** [comment]
@fwallb the essay is ready for your feedback. Give us your thoughts and we will do our best to implement them in time! 🛺

**Comment by fwallb** [comment]
Hi @ayubatif & @agriad, 

This is an essay on an interesting subject that is very relevant to the course. Choosing the right tool that fits the needs of your project is important. The cost aspect is also essential since setting up a tool, and then having to choose another tool could imply loss of money. The essay has a great layout and figures that are visually pleasing. It is great that it is written in LaTex, which gives it a professional feel.

It is very ambitious to compare three types of software tools and three providers in each category. Doing this in less than 2000 words is a real challenge and gives approximately 200 words per provider. In your case with only ~1500, words due to many references, there are only 166 words per provider. This means that there is no space to go in depth to really understand their differences. Maybe it would have been better to focus on one tool, i.e CI tools, cloud providers or code repositories, to be able to describe them fully. What is the suggestion to improve and give the reader more information? Some of the facts in the text could be converted to a table or figure (like figure 2), to give more room for text. Maybe a few of the references could be removed to give more room for text?

The 2text-flow2 missing in some parts of the essay, and there are some grammar and spelling mistakes. A proofread and a run through Grammarly would improve the essay.

There are in some places, big jumps between sentences which makes the text hard to follow and understand. The language is somewhat informal and the tense used in some parts of the text is not fitting for a formal report. For example _"This article will be looking at..."_ is future tense, but the article is finished, it would be better to use _"This article looks at..."_.

Further, the use of first-person pronouns is advised against in formal writing. For example _"...we will be comparing..."_, the sentence could be changed to _"Within continuous integration, Jenkins, Travis CI, and Circle CI highlighting tool effectiveness and use cases __will be compared__."

Some parts of the text, mainly the second section _Different CI tools_, consists of many very short sentences. Longer sentences are preferred in formal texts.

There are different usages of references throughout the text, especially when two references are used. You use [x][y], [x] [y], [x]. [y] and [x], [y] - would be better if it was consistent.

The title suggests that there is some comparison made in the essay, but it mainly consists of facts. It would have benefitted from a discussion at the end to really understand and get the author's perspective on the different tools.

There are many references, possibly too many, so that the substantial part of the article might be considered too short. 1558 words are included in the main part of the document and the rest are references. A large portion of the sentences are ended with a reference(s), which could give a messy feel.

### 1 Introduction
Good that the introduction is short and gives a good overview of what a reader can expect to learn from the essay.

#### Potential improvements
- The last sentence is a bit bold - maybe change _"gain confidence with such difficult decisions"_ to _"have more information when making such decisions"_.

### 2 Different CI tools
- I really like the figure! On a second look - it looks somewhat aggressive with Travis holding those sharp objects...
- The aspects that are used for comparison are interesting and seem like good evaluation factors.

#### Potential improvements
- In the first paragraph there is a _"but"_ and then _"only"_ which indicates some downgrading of the essay from the authors' point of view. Might be better to rephrase to something like _"the three tools compared in this section are; Travis CI..."_
- The text in the third paragraph, starting with _"Travis CI was known for..."_, feels like more of a list than a text when reading. The sentences do not work with each other. Maybe it is better to convert this part to a list of "Travis facts"? Or rewrite? After further thought, there might better to have a table with the different aspects, like:
|           | Jenkins | TravisCI | Circle CI   |
|---------- | --------|----------|-------------|
|<aspect 1> |         |          |             |
|<aspect 2> |         |          |             |

- There is a gap between the forth and the fifth paragraph, seems like the paragraphs could be merged.
- There is limited information about Travis CI, only 5 sentences, which is a little thin.

#### Grammar and spelling
- One type of __tool__
- There are many different software that __provides__
- Open source software -> open-source software
- As OSS... -> As an OSS...
-  ... customize their Jenkins setup to fit their __needs__.
- Open source community -> open-source community
- Code base -> codebase
- ...using others __requires__ tinkering

### 3 Different cloud providers
Good that there are a few introduction sentences in the beginning. Great that you have a figure that gives a good overview - more of those! It is easy to read and gives a lot of information quickly without using too many words.

#### Potential improvements
- The second paragraph is only one sentence long, and could maybe be put together with the third paragraph.

#### Grammar and spelling
- GCP’s cloud resources __encompasses__...
- High profile -> high-profile
- Pay-as-yo-go is spelled differently throughout
- Cost effective -> cost-effective
- google -> Google

### 4 Different code repositories
This part has a good introduction!

#### Potential improvements
- The graph in figure 2 lacks y-axis label.
- _"...and Dow Jones are some of the customers that GitHub has..."_ could be rewritten as _"...and Dow Jones are some of GitHub's customers."_ to improve clarity.
- _"...in terms of __the__ number of users..."_
- The second to last and last paragraphs are separated but it seems like they could be merged.

#### Grammar and spelling
- ...repository hosting company __whose__ users
-  management oriented features -> management-oriented features
- kanban -> Kanban

### 5 Conclusion
The conclusion is not really a conclusion, maybe it is more of a summary. Missing is some sort of ending and conclusion of the findings. [This](https://www.academic-englishuk.com/conclusions) might give some inspiration.

### References
The references are many, maybe too many? They seem to be correctly formated which is great!

**Comment by ayubatif** [comment]
Hello @fwallb. Thank you for the detailed feedback! 🏹 

## Topic evaluation 
It was refreshing to hear your take on our essay. It is true that there was a large scope for the essay within the word limit, so we settled for giving the reader a quick summary of each tool, notable users, particular use-case niches, and unique features providing a competitive edge. It was smart to note that we should more accurately depict this in the introduction and it also ties into our short albeit concise conclusion which reiterates the most distinct usage niches.  You ask how we can describe the tools more fully, but that would lead to an unjust comparison as the word limit does not allow such freedom for describing each tool at leisure, we instead hope for our citations to lead to further reading to provide such descriptions. 

## Structure
This leads to an unorthodox text flow, but we thought it the best way to use the word count constraint with our topic. We actually had more citations in our bib file prior to reaching the word limit, and the reason for so many references is that we make many claims and references to facts and functionality for the devops platforms and it only felt natural to provide the source to all those claims, this helps assure the soundness and factual nature of the essay. We understand the need for original work, and the citations being there are sort of a necessity given the topic we address. Indeed, this is an essay of largely facts highlighted to guide the reader to realize certain aspects of all these tools to make educated decisions.

## Writing
Thank you for the corrections you have provided us. Proofreading really is vital and we missed a few things. We have also taken most of your comments for improvement and implemented them

### Change Log
- Fixed chart y-axis
- Fixed tense, first-person, and references.
- Added Introduction suggested improvements 
- Added Different CI tools suggested  improvements 
- Added Different cloud services suggested improvements 
- Added Different code repositories suggested improvements

**Comment by SophieHYe** [comment]
Thanks for the final essay submission. I am now merging your PR and we will grade your work based on the final version.
